File: ios/Classes/Flutter/Loadable/LoadableNativeObject.swift
//
// Created by Valentin Grigorean on 25.07.2022.
//

import Foundation
import ArcGIS

class LoadableNativeObject: NativeHandler {
    let loadable: AGSLoadable
    var loadStatus: AGSLoadStatus

    init(loadable: AGSLoadable) {
        self.loadable = loadable
        loadStatus = loadable.loadStatus
    }

    var messageSink: NativeMessageSink?

    func dispose() {
        messageSink = nil
    }

    func onMethodCall(method: String, arguments: Any?, result: @escaping FlutterResult) -> Bool {
        switch (method) {
        case "getLoadStatus":
            result(loadable.loadStatus.rawValue)
            return true
        case "getLoadError":
            result(loadable.loadError?.toJSON())
            return true
        case "cancelLoad":
            loadable.cancelLoad()
            result(nil)
            return true
        case "loadAsync":
            loadable.load { [weak self] (error) in
                guard let self = self else {
                    return
                }
                self.onLoadCallback(error: error, result: result)
            }
            return true
        case "retryLoadAsync":
            loadable.retryLoad { [weak self] (error) in
                guard let self = self else {
                    return
                }
                self.onLoadCallback(error: error, result: result)
            }
            return true
        default:
            return false
        }
    }


    private func onLoadCallback(error: Error?, result: @escaping FlutterResult) {
        let newStatus = loadable.loadStatus
        if newStatus != loadStatus {
            loadStatus = newStatus
            messageSink?.send(method: "loadStatusChanged", arguments: loadStatus.rawValue)
        }
        result(nil)
    }

}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/loadable/LoadableNativeObject.java
package com.valentingrigorean.arcgis_maps_flutter.loadable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.esri.arcgisruntime.loadable.LoadStatusChangedEvent;
import com.esri.arcgisruntime.loadable.LoadStatusChangedListener;
import com.esri.arcgisruntime.loadable.Loadable;
import com.valentingrigorean.arcgis_maps_flutter.flutter.ArcgisNativeObjectController;
import com.valentingrigorean.arcgis_maps_flutter.Convert;
import com.valentingrigorean.arcgis_maps_flutter.flutter.NativeMessageSink;

import io.flutter.plugin.common.MethodChannel;

public class LoadableNativeObject implements ArcgisNativeObjectController.NativeHandler, LoadStatusChangedListener {

    private final Loadable loadable;
    private NativeMessageSink messageSink;
    private boolean disposed = false;

    public LoadableNativeObject(Loadable loadable) {
        this.loadable = loadable;
        this.loadable.addLoadStatusChangedListener(this);
    }

    @Override
    public void dispose() {
        if (disposed) {
            return;
        }
        disposed = true;
        loadable.removeLoadStatusChangedListener(this);
    }

    @Override
    public void setMessageSink(@Nullable NativeMessageSink messageSink) {
        this.messageSink = messageSink;
    }

    @Override
    public boolean onMethodCall(@NonNull String method, @Nullable Object args, @NonNull MethodChannel.Result result) {
        switch (method) {
            case "loadable#getLoadStatus":
                result.success(loadable.getLoadStatus().ordinal());
                return true;
            case "loadable#getLoadError":
                result.success(Convert.arcGISRuntimeExceptionToJson(loadable.getLoadError()));
                return true;
            case "loadable#cancelLoad":
                loadable.cancelLoad();
                result.success(null);
                return true;
            case "loadable#loadAsync":
                loadable.loadAsync();
                loadable.addDoneLoadingListener(new DoneListener(result));
                return true;
            case "loadable#retryLoadAsync":
                loadable.retryLoadAsync();
                loadable.addDoneLoadingListener(new DoneListener(result));
                return true;
        }
        return false;
    }

    @Override
    public void loadStatusChanged(LoadStatusChangedEvent loadStatusChangedEvent) {
        final NativeMessageSink messageSink = this.messageSink;
        if (messageSink == null) {
            return;
        }
        messageSink.send("loadable#loadStatusChanged", loadStatusChangedEvent.getNewLoadStatus().ordinal());
    }

    private class DoneListener implements Runnable {
        private final MethodChannel.Result result;

        private DoneListener(MethodChannel.Result result) {
            this.result = result;
        }

        @Override
        public void run() {
            loadable.removeDoneLoadingListener(this);
            if (!disposed) {
                result.success(null);
            }
        }
    }
}
